### PR TITLE
[CUDA] Embedding kernel and testing with pytorch golden values

### PR DESCRIPTION
**Pull request by feiwang3311**

### CODE REVIEW

File: src/main/resources/headers/scanner_header.h [hunk 1]
```diff
@@ -101,24 +109,18 @@ void scan_int_rank(const char *filename, int rank, int *array, int size) {
...
   snprintf(buf, 2000, "%s_rank_%d.data", filename, rank);
-  if (check_float_array(buf, gold, check, size))
-    fprintf(stdout, "Checking value of %s passed!\n", buf);
-  else
-    fprintf(stderr, "Checking value of %s failed!\n", buf);
+  check_float_array(buf, check, size);
 }
 
-void check_int_array_rank(const char *filename, int rank, int *gold, int *check, int size) {
+void check_int_array_rank(const char *filename, int rank, int *check, int size) {
```
Comment (feiwang3311) [inline src/main/resources/headers/scanner_header.h:119]: signature made simpler

File: src/main/resources/headers/scanner_header.h [hunk 2]
```diff
@@ -101,24 +109,18 @@ void scan_int_rank(const char *filename, int rank, int *array, int size) {
   scan_int(buf, array, size);
 }
 
-void check_float_array_rank(const char *filename, int rank, float *gold, float *check, int size) {
+void check_float_array_rank(const char *filename, int rank, float *check, int size) {
```
Comment (feiwang3311) [inline src/main/resources/headers/scanner_header.h:112]: signature made simpler

File: src/main/scala/lms/thirdparty/cuda/cuda.scala [hunk 3]
```diff
@@ -943,6 +943,26 @@ trait CudaOps extends Dsl with StackArrayOps with SizeTOps with CLibs with CudaF
   }
 }
 
+trait CudaLibs extends CudaOps {
```
Comment (feiwang3311) [inline src/main/scala/lms/thirdparty/cuda/cuda.scala:946]: I am wondering if it is a good idea to move Cuda Kernel implementations to this separate trait called CudaLibs?

File: src/main/scala/lms/thirdparty/cuda/cuda.scala [hunk 4]
```diff
@@ -943,6 +943,26 @@ trait CudaOps extends Dsl with StackArrayOps with SizeTOps with CLibs with CudaF
...
+  def cudaEmbedding[T:Numeric:Manifest](implicit __pos: SourceContext) = cudaGlobalFun {
+    (embedding: Rep[Array[T]], indices: Rep[Array[Int]], output: Rep[Array[T]], embed_size: Rep[Int]) => {
+      generate_comment("this is cuda embedding kernel.")
+      generate_comment("arg0: 2D embedding table: <n_embedding x embed_size>")
+      generate_comment("arg1: 1D indices: <indices_size>")
+      generate_comment("arg2: 2D output: <indices_size x embed_size>")
+      generate_comment("arg3: embed_size")
+      generate_comment("invocation assumption: <<<dim3(a,1,1), dim3(indices_size,1,1)>>> where a <= embed_size")
+      generate_comment("each thread block handles one embedding vector")
```
Comment (feiwang3311) [inline src/main/scala/lms/thirdparty/cuda/cuda.scala:955]: good habit to give a list of comments that will show up in the generated kernel and explain the calling conventions

File: src/test/scala/lms/thirdparty/test_cuda.scala [hunk 5]
```diff
@@ -236,12 +241,41 @@ class CudaTest extends TutorialFunSuite {
...
+  test("embedding") {
+    val driver = new DslDriverCCudeScan[Int, Unit] {
+
+      @virtualize
+      def snippet(arg: Rep[Int]) = {
+        val n_embeddings = 20
+        val embed_size = 60
+
+        val embedding = NewArray[Float](n_embeddings * embed_size)
+        scanFile[Float]("golden/embedding/embedding.data", embedding, n_embeddings * embed_size)
```
Comment (feiwang3311) [inline src/test/scala/lms/thirdparty/test_cuda.scala:258]: unfortunately we have to use some naming conventions in scanFile and checkFile

File: src/main/scala/lms/thirdparty/cuda/cuda.scala (hunk 3 above)
Comment (supunab) [reply src/main/scala/lms/thirdparty/cuda/cuda.scala:946]: yeah, this makes sense. I'll do the same for softmax, softmaxGrad after we merge this PR

File: src/test/scala/lms/transformation/test_distributed_tensor/conv.py [hunk 6]
```diff
@@ -8,15 +8,15 @@ def generate_data(lms_clean_root: str):
     torch.manual_seed(0)
 
     # model
-    input = torch.randn(temp, (2, 1, 9, 9))
+    input = torch.randn(2, 1, 9, 9)
```
[the remark this answers is not in the document]
Comment (luke-jiang) [reply src/test/scala/lms/transformation/test_distributed_tensor/conv.py:11]: I probably forgot to update the .py functions in the branch, I updated them on the server. I rechecked conv with this updated script and it passes the test.

File: src/test/scala/lms/transformation/test_distributed_tensor/avgpool.py [hunk 7]
```diff
@@ -18,7 +18,7 @@ def generate_data(lms_clean_root: str):
     loss.sum().backward()
 
     # printer
-    printer = get_printer(lms_clean_root, test_name = "maxpool")
+    printer = get_printer(lms_clean_root, test_name = "avgpool")
```
[the remark this answers is not in the document]
Comment (luke-jiang) [reply src/test/scala/lms/transformation/test_distributed_tensor/avgpool.py:21]: it is "avgpool" on my server. I may forgot updating .py files before merging. Sorry about that!

File: src/main/scala/lms/thirdparty/cuda/cuda.scala (hunk 4 above)
Comment (luke-jiang) [reply src/main/scala/lms/thirdparty/cuda/cuda.scala:955]: will do the same for other kernels.

**Review by luke-jiang — APPROVED** [review]
